Intermittent test failure - test_subprocess_print

Seen on Jenkins on a Linux build:
https://jenkins.jupyter.org/job/jack-of-none/142/console

OK when the build was repeated.

```
======================================================================
FAIL: printing from forked mp.Process
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/var/lib/jenkins/shiningpanda/jobs/92f5a430/virtualenvs/d41d8cd9/lib/python3.4/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/var/lib/jenkins/shiningpanda/jobs/92f5a430/virtualenvs/d41d8cd9/lib/python3.4/site-packages/IPython/external/decorators/_decorators.py", line 222, in knownfailer
    return f(*args, **kwargs)
  File "/var/lib/jenkins/shiningpanda/jobs/92f5a430/virtualenvs/d41d8cd9/lib/python3.4/site-packages/ipykernel/tests/test_kernel.py", line 83, in test_subprocess_print
    nt.assert_equal(stdout.count("hello"), np, stdout)
nose.proxy.AssertionError: 4 != 5 : hello 0
hello 2
hello 4
hello 1

-------------------- >> begin captured logging << --------------------
root: DEBUG: Found kernel python3 in /tmp/tmpqp6_2x_7/.local/share/jupyter/kernels
root: DEBUG: Starting kernel: ['/var/lib/jenkins/shiningpanda/jobs/92f5a430/virtualenvs/d41d8cd9/bin/python3.4', '-m', 'ipykernel', '-f', '/tmp/tmpwheu19w3.json']
root: DEBUG: Connecting to: tcp://127.0.0.1:58481
root: DEBUG: connecting shell channel to tcp://127.0.0.1:33469
root: DEBUG: Connecting to: tcp://127.0.0.1:33469
root: DEBUG: connecting iopub channel to tcp://127.0.0.1:59032
root: DEBUG: Connecting to: tcp://127.0.0.1:59032
root: DEBUG: connecting stdin channel to tcp://127.0.0.1:38519
root: DEBUG: Connecting to: tcp://127.0.0.1:38519
root: DEBUG: connecting heartbeat channel to tcp://127.0.0.1:50245
--------------------- >> end captured logging << ---------------------
```

--- a/ipykernel/iostream.py
+++ b/ipykernel/iostream.py
@@ -48,7 +48,6 @@ class IOPubThread(object):
         self.socket = socket
         self.background_socket = BackgroundSocket(self)
         self._master_pid = os.getpid()
-        self._pipe_pid = os.getpid()
         self._pipe_flag = pipe
         self.io_loop = IOLoop()
         if pipe:
@@ -94,24 +93,19 @@ class IOPubThread(object):
     def _setup_pipe_out(self):
         # must be new context after fork
         ctx = zmq.Context()
-        self._pipe_pid = os.getpid()
-        self._pipe_out = ctx.socket(zmq.PUSH)
-        self._pipe_out.connect("tcp://127.0.0.1:%i" % self._pipe_port)
+        pipe_out = ctx.socket(zmq.PUSH)
+        pipe_out.linger = 3000 # 3s timeout for pipe_out sends before discarding the message
+        pipe_out.connect("tcp://127.0.0.1:%i" % self._pipe_port)
+        return ctx, pipe_out
 
     def _is_master_process(self):
         return os.getpid() == self._master_pid
-
-    def _have_pipe_out(self):
-        return os.getpid() == self._pipe_pid
 
     def _check_mp_mode(self):
         """check for forks, and switch to zmq pipeline if necessary"""
         if not self._pipe_flag or self._is_master_process():
             return MASTER
         else:
-            if not self._have_pipe_out():
-                # setup a new out pipe
-                self._setup_pipe_out()
             return CHILD
     
     def start(self):
@@ -156,14 +150,12 @@ class IOPubThread(object):
             self.socket.send_multipart(msg, *args, **kwargs)
         else:
             # we are a child, pipe to master
-            kwargs['copy'] = False
-            kwargs['track'] = True
-            tracker = self._pipe_out.send_multipart([self._pipe_uuid] + msg, *args, **kwargs)
-            try:
-                tracker.wait(1)
-            except Exception as e:
-                print("Failed to send: %s" % e, file=sys.__stderr__)
-                pass
+            # new context/socket for every pipe-out
+            # since forks don't teardown politely, use ctx.term to ensure send has completed
+            ctx, pipe_out = self._setup_pipe_out()
+            pipe_out.send_multipart([self._pipe_uuid] + msg, *args, **kwargs)
+            pipe_out.close()
+            ctx.term()
 
 
 class BackgroundSocket(object):

--- a/ipykernel/tests/test_kernel.py
+++ b/ipykernel/tests/test_kernel.py
@@ -75,10 +75,6 @@ def test_subprocess_print():
             "time.sleep(0.5),"
         ])
 
-        expected = '\n'.join([
-            "hello %s" % i for i in range(np)
-        ]) + '\n'
-
         msg_id, content = execute(kc=kc, code=code)
         stdout, stderr = assemble_output(iopub)
         nt.assert_equal(stdout.count("hello"), np, stdout)

